### PR TITLE
Add Nested Function Support in SELECT Clause

### PR DESCRIPTION
**Pull request by forestmvey**
### Description
Syntax: `nested( [field} | [field,path] )`

Nested function allows the query of nested types of an index. The `path` parameter is determined dynamically if not provided by the user, and the output query should be identical assuming the user input the correct parameter values. The `condition` parameter is not supported when the nested function is used in the `SELECT` clause. Nested types structure is flattened making the full path of an object the key, and the object it refers to the value.
```
   Sample input=
   keys = ['comments.likes']
   row = comments: {
   likes: 2
   }
   output =
   flattenedRow = {comment.likes: 2}
```

#### Simple nested query, array of `message.info` is flattened into two rows.
`SELECT nested(message.info) FROM nested_objects;`
#### After flattening notice `comment.data` has repeating `ab`
`SELECT nested(message.info), nested(comment.data) FROM nested_objects;`
#### someField is of type object
`SELECT nested(message.info), someField FROM nested_objects;`

### To Do

- nested function nested in other function call (Not currently supported in legacy)

### Not Supported

- Using `condition` parameter in the `SELECT` clause.

### Changes from Legacy Functionality

- Nested function does not support `condition` parameter in SELECT clause.
- Updated `partiql.rst` to handle object arrays same as V2 engine. 

 
### Issues Resolved
- [1111](https://github.com/opensearch-project/sql/issues/1111)
- [1277](https://github.com/opensearch-project/sql/issues/1277)
- [1489](https://github.com/opensearch-project/sql/issues/1489)
 
### Check List
- [x] New functionality includes testing.
  - [x] All tests pass, including unit test, integration test and doctest
- [x] New functionality has been documented.
  - [x] New functionality has javadoc added
  - [ ] New functionality has user manual doc added
- [x] Commits are signed per the DCO using --signoff 

By submitting this pull request, I confirm that my contribution is made under the terms of the Apache 2.0 license.
For more information on following Developer Certificate of Origin and signing off your commits, please check [here](https://github.com/opensearch-project/OpenSearch/blob/main/CONTRIBUTING.md#developer-certificate-of-origin).

### CODE REVIEW

File: sql/src/main/java/org/opensearch/sql/sql/parser/AstExpressionBuilder.java [hunk 1]
```diff
@@ -151,6 +156,30 @@ public UnresolvedExpression visitScalarFunctionCall(ScalarFunctionCallContext ct
     return buildFunction(ctx.scalarFunctionName().getText(), ctx.functionArgs().functionArg());
   }
 
+  @Override
+  public UnresolvedExpression visitNestedFunctionCall(NestedFunctionCallContext ctx) {
+    List<UnresolvedExpression> args = new ArrayList();
+    boolean conditionSet = false;
+
+    if (ctx.nestedFunction().nestedField() != null) {
+      args.add(new QualifiedName(ctx.nestedFunction().nestedField().getText()));
```
Comment (Yury-Fridlyand) [inline sql/src/main/java/org/opensearch/sql/sql/parser/AstExpressionBuilder.java:165]: Don't you need to `visitQualifiedName` here?

File: sql/src/main/java/org/opensearch/sql/sql/parser/AstUnnestBuilder.java [hunk 2]
```diff
@@ -0,0 +1,58 @@
...
+   */
+  private final QuerySpecification querySpec;
+
+  @Override
+  public UnresolvedPlan visit(ParseTree selectClause) {
+    Unnest unnest = null;
+    for (UnresolvedExpression item : querySpec.getSelectItems()) {
+      // Assume is nested as that's the only function currently
+      // supported as alias in SELECT clause from AstBuilder
+      if (item instanceof Alias) {
```
Comment (Yury-Fridlyand) [inline sql/src/main/java/org/opensearch/sql/sql/parser/AstUnnestBuilder.java:33]: The loop lap skipped if `item` is not an `Alias`. And `visit` returns `null` if all `item`s aren't `Alias`es. Just to confirm that it is corrent.

File: integ-test/src/test/java/org/opensearch/sql/sql/NestedIT.java [hunk 3]
```diff
@@ -0,0 +1,116 @@
...
+
+    assertEquals(6, result.getInt("total"));
+    verifyDataRows(result,
+        rows("e"),
+        rows("f"),
+        rows("g"),
+        rows("h"),
+        rows("p"),
+        rows("yy"));
+  }
```
Comment (Yury-Fridlyand) [inline integ-test/src/test/java/org/opensearch/sql/sql/NestedIT.java:115]: Add a test with `nested` with condition
Add test(s) with incorrect input (missing subfield or path is not of `nested` type)

**Review by Yury-Fridlyand — COMMENTED** [review]
Cool!
Doctests? Demo? Nested party?

File: core/src/main/java/org/opensearch/sql/expression/nested/NestedFunctions.java [hunk 4]
```diff
@@ -0,0 +1,144 @@
...
+
+import lombok.experimental.UtilityClass;
+import org.opensearch.sql.data.model.ExprValue;
+import org.opensearch.sql.data.model.ExprValueUtils;
+import org.opensearch.sql.expression.function.BuiltinFunctionName;
+import org.opensearch.sql.expression.function.BuiltinFunctionRepository;
+import org.opensearch.sql.expression.function.DefaultFunctionResolver;
+import org.opensearch.sql.expression.function.FunctionDSL;
+
+@UtilityClass
```
Comment (Yury-Fridlyand) [inline core/src/main/java/org/opensearch/sql/expression/nested/NestedFunctions.java:25]: This could be troublesome: https://github.com/opensearch-project/sql/pull/1045#discussion_r1023132711

File: core/src/main/java/org/opensearch/sql/expression/nested/NestedFunctions.java [hunk 5]
```diff
@@ -0,0 +1,144 @@
...
+            }, BOOLEAN, BOOLEAN, STRUCT),
+        FunctionDSL.impl(
+            (v1, v2) -> {
+              return nullMissingFirstArgOnlyHandling(v1);
+            }, STRUCT, STRUCT, STRUCT),
+        FunctionDSL.impl(
+            (v1, v2) -> {
+              return nullMissingFirstArgOnlyHandling(v1);
+            }, ARRAY, ARRAY, STRUCT));
+  }
```
[the remark this answers is not in the document]
Comment (forestmvey) [reply core/src/main/java/org/opensearch/sql/expression/nested/NestedFunctions.java:129]: Yes that is correct!

File: opensearch/src/main/java/org/opensearch/sql/opensearch/storage/scan/OpenSearchIndexScanQueryBuilder.java [hunk 6]
```diff
@@ -99,6 +100,13 @@ public boolean pushDownHighlight(LogicalHighlight highlight) {
     return true;
   }
 
+  @Override
+  public boolean pushDownNested(LogicalNested nested) {
+    indexScan.getRequestBuilder().pushDownNested(nested.getFields());
+    // Return false intentionally to keep the original nested operator
+    return false;
```
[the remark this answers is not in the document]
Comment (forestmvey) [reply opensearch/src/main/java/org/opensearch/sql/opensearch/storage/scan/OpenSearchIndexScanQueryBuilder.java:107]: To keep the PhysicalPlan for `UnnestOperator` to iterate over the input from OS. I am having issues with this currently because I want the `ProjectOperator` pushed down while also keeping the PhysicalPlan for the `UnnestOperator` to do processing on the input from OS. After pushDown if we return true then that portion of the  PhysicalPlan is removed from the list.

File: sql/src/main/java/org/opensearch/sql/sql/parser/AstUnnestBuilder.java (hunk 2 above)
Comment (forestmvey) [reply sql/src/main/java/org/opensearch/sql/sql/parser/AstUnnestBuilder.java:33]: Yes, this may change dependent on if I can achieve pushDown for the ProjectOperator or not though.

File: core/src/main/java/org/opensearch/sql/expression/nested/NestedFunctions.java [hunk 7]
```diff
@@ -0,0 +1,144 @@
...
+
+  private static DefaultFunctionResolver nested() {
+    return FunctionDSL.define(BuiltinFunctionName.NESTED.getName(),
+        FunctionDSL.impl(
+            FunctionDSL.nullMissingHandling(
+                (v1) -> v1
+            ), STRING, STRING),
+        FunctionDSL.impl(
+            FunctionDSL.nullMissingHandling(
```
[the remark this answers is not in the document]
Comment (forestmvey) [reply core/src/main/java/org/opensearch/sql/expression/nested/NestedFunctions.java]: Fixed in [6fd757416c848d6e2d11f727791f8ebdd1cc9f99](https://github.com/Bit-Quill/opensearch-project-sql/pull/243/commits/6fd757416c848d6e2d11f727791f8ebdd1cc9f99)

File: core/src/main/java/org/opensearch/sql/expression/nested/NestedFunctions.java [hunk 8]
```diff
@@ -0,0 +1,144 @@
...
+  public static void register(BuiltinFunctionRepository repository) {
+    repository.register(nested());
+  }
+
+  private static DefaultFunctionResolver nested() {
+    return FunctionDSL.define(BuiltinFunctionName.NESTED.getName(),
+        FunctionDSL.impl(
+            FunctionDSL.nullMissingHandling(
+                (v1) -> v1
+            ), STRING, STRING),
```
[the remark this answers is not in the document]
Comment (forestmvey) [reply core/src/main/java/org/opensearch/sql/expression/nested/NestedFunctions.java]: Fixed in [6fd757416c848d6e2d11f727791f8ebdd1cc9f99](https://github.com/Bit-Quill/opensearch-project-sql/pull/243/commits/6fd757416c848d6e2d11f727791f8ebdd1cc9f99)

File: core/src/main/java/org/opensearch/sql/planner/physical/UnnestOperator.java [hunk 9]
```diff
@@ -0,0 +1,180 @@
...
+import org.opensearch.sql.data.model.ExprTupleValue;
+import org.opensearch.sql.data.model.ExprValue;
+import org.opensearch.sql.expression.ReferenceExpression;
+
+@EqualsAndHashCode(callSuper = false)
+public class UnnestOperator extends PhysicalPlan {
+  @Getter
+  private final PhysicalPlan input;
+  @Getter
+  private final Set<String> fields; // Needs to be a Set to match legacy implementation
```
[the remark this answers is not in the document]
Comment (forestmvey) [reply core/src/main/java/org/opensearch/sql/planner/physical/UnnestOperator.java:31]: Order matters and must match legacy or we must alter legacy tests.

File: core/src/main/java/org/opensearch/sql/analysis/NestedAnalyzer.java [hunk 10]
```diff
@@ -0,0 +1,65 @@
...
+  private final LogicalPlan child;
+
+  public LogicalPlan analyze(UnresolvedExpression projectItem, AnalysisContext context) {
+    LogicalPlan nested = projectItem.accept(this, context);
+    return (nested == null) ? child : nested;
+  }
+
+  @Override
+  public LogicalPlan visitAlias(Alias node, AnalysisContext context) {
+    if (node.getDelegated() instanceof Function
```
Comment (GumpacG) [inline core/src/main/java/org/opensearch/sql/analysis/NestedAnalyzer.java:39]: Would overriding the `visitFunction` work here?
Something like:
```
  @Override
  public LogicalPlan visitAlias(Alias node, AnalysisContext context) {
    return node.getDelegated().accept(this, context);
  }

  @Override
  public LogicalPlan visitFunction(Function node, AnalysisContext context) {
    if (node.getFuncName().equalsIgnoreCase("nested")) {

      List<UnresolvedExpression> expressions = node.getFuncArgs();
      ReferenceExpression nestedField =
          (ReferenceExpression)expressionAnalyzer.analyze(expressions.get(0), context);
      Map<String, ReferenceExpression> args;
      if (expressions.size() == 2) {
        args = Map.of(
            "field", nestedField,
            "path", (ReferenceExpression)expressionAnalyzer.analyze(expressions.get(1), context)
        );
      } else {
        args = Map.of(
            "field", (ReferenceExpression)expressionAnalyzer.analyze(expressions.get(0), context),
            "path", generatePath(nestedField.toString())
        );
      }
      return new LogicalNested(child, List.of(args), namedExpressions);
    }
    return null;
  }

**Comment by GumpacG** [comment]
Nested with date type subfields are supported however it returns the wrong type. 
**Legacy:**
Screenshot 2023-03-20 at 3 32 13 PM
**V2:**
Screenshot 2023-03-20 at 3 32 28 PM

File: sql/src/main/antlr/OpenSearchSQLParser.g4 [hunk 11]
```diff
@@ -631,6 +645,14 @@ ident
...
 
+
+nestedIdent
+    : ID
+    | BACKTICK_QUOTE_ID
+    | keywordsCanBeId
+    | scalarFunctionName
+    ;
+
 keywordsCanBeId
```
Comment (GumpacG) [inline sql/src/main/antlr/OpenSearchSQLParser.g4:656]: Add `NESTED` here to allow tables with that name

File: sql/src/main/antlr/OpenSearchSQLParser.g4 [hunk 12]
```diff
@@ -323,6 +324,19 @@ getFormatType
     | TIMESTAMP
     ;
 
+nestedFunction
+    : NESTED LR_BRACKET
+    (nestedField | nestedField COMMA nestedPath | nestedPath COMMA expression)
```
Comment (GumpacG) [inline sql/src/main/antlr/OpenSearchSQLParser.g4:329]: Are the last two cases relevant for SELECT? If not, maybe remove for now until WHERE implementation?

File: core/src/main/java/org/opensearch/sql/analysis/NestedAnalyzer.java (hunk 10 above)
Comment (forestmvey) [reply core/src/main/java/org/opensearch/sql/analysis/NestedAnalyzer.java:39]: Nice, I think this is a nicer implementation!

File: integ-test/src/test/java/org/opensearch/sql/sql/NestedIT.java (hunk 3 above)
Comment (forestmvey) [reply integ-test/src/test/java/org/opensearch/sql/sql/NestedIT.java:116]: > Nested with date type subfields are supported however it returns the wrong type. Maybe date types need to be included in NestedFunction.java. **Legacy:** Screenshot 2023-03-20 at 3 32 13 PM **V2:** Screenshot 2023-03-20 at 3 32 28 PM

Appears to be a bug with not only nested types!

File: sql/src/main/java/org/opensearch/sql/sql/parser/AstExpressionBuilder.java (hunk 1 above)
Comment (forestmvey) [reply sql/src/main/java/org/opensearch/sql/sql/parser/AstExpressionBuilder.java]: Fixed in [4e4f4a43d26a3420d401878801154096a6afe728](https://github.com/Bit-Quill/opensearch-project-sql/pull/243/commits/4e4f4a43d26a3420d401878801154096a6afe728)

File: core/src/test/java/org/opensearch/sql/expression/nested/NestedFunctionTest.java [hunk 13]
```diff
@@ -0,0 +1,112 @@
...
+      "message.long_value", longValue(1L),
+      "message.float_value", floatValue(1f),
+      "message.double_value", doubleValue(1d),
+      "message.boolean_value", booleanValue(true),
+      "message.string_value", stringValue("str"),
+      "message.struct_value", tupleValue(ImmutableMap.of("str", 1)),
+      "message.array_value", collectionValue(ImmutableList.of(1))
+  );
+
+  protected static Environment<Expression, ExprValue> env() {
```
[the remark this answers is not in the document]
Comment (forestmvey) [reply core/src/test/java/org/opensearch/sql/expression/nested/NestedFunctionTest.java:47]: `nested.valueOf()` gets nullPointerException without this!

File: core/src/test/java/org/opensearch/sql/planner/DefaultImplementorTest.java [hunk 14]
```diff
@@ -97,61 +99,69 @@ public void visitShouldReturnDefaultPhysicalOperator() {
...
-                                        mappings),
-                                    exclude),
-                                newEvalField),
-                            sortField),
-                        CommandType.TOP,
-                        topByExprs,
-                        rareTopNField),
-                    dedupeField),
-                limit,
+            nested(
```
[the remark this answers is not in the document]
Comment (forestmvey) [reply core/src/test/java/org/opensearch/sql/planner/DefaultImplementorTest.java:123]: But then we get less memes...

File: core/src/test/java/org/opensearch/sql/expression/nested/NestedFunctionTest.java [hunk 15]
```diff
@@ -0,0 +1,112 @@
...
+    );
+  }
+
+  /**
+   * Test single parameter nested function parameters.
+   * @param dataMap : Map of data to validate against.
+   */
+  @ParameterizedTest
+  @MethodSource("generateValidData")
+  public void test_single_param_nested(Map<String, ExprValue> dataMap) {
```
[the remark this answers is not in the document]
Comment (forestmvey) [reply core/src/test/java/org/opensearch/sql/expression/nested/NestedFunctionTest.java]: Fixed: [22e15537948b759963124613e03d616602ec7811](https://github.com/Bit-Quill/opensearch-project-sql/pull/243/commits/22e15537948b759963124613e03d616602ec7811)

File: core/src/main/java/org/opensearch/sql/planner/physical/UnnestOperator.java [hunk 16]
```diff
@@ -0,0 +1,252 @@
...
+   * <p>Return:
+   * flattenedRow = {comment.likes: 2}
+   *
+   * @param nestedField : Field to query in row.
+   * @param row : Row returned from OS.
+   * @param prevList : List of previous nested calls.
+   * @param supportArrays : When false we do not need to execute a cross join.
+   * @return : List of nested select items or cartesian product of nested calls.
+   */
+  @SuppressWarnings("unchecked")
```
Comment (acarbonetto) [inline core/src/main/java/org/opensearch/sql/planner/physical/UnnestOperator.java:157]: can we avoid suppressing?

File: core/src/main/java/org/opensearch/sql/planner/physical/UnnestOperator.java [hunk 17]
```diff
@@ -0,0 +1,252 @@
...
+  private List<Map<String, ExprValue>> flatten(
+      String nestedField,
+      ExprValue row,
+      List<Map<String,
+      ExprValue>> prevList,
+      boolean supportArrays
```
Comment (acarbonetto) [inline core/src/main/java/org/opensearch/sql/planner/physical/UnnestOperator.java:163]: can we make supportArrays == false a separate function?  Feels like these have different behaviours anyways

File: core/src/test/java/org/opensearch/sql/analysis/AnalyzerTest.java [hunk 18]
```diff
@@ -367,6 +370,33 @@ public void project_source() {
             AstDSL.alias("double_value", AstDSL.field("double_value"))));
   }
 
+  @Test
```
[the remark this answers is not in the document]
Comment (acarbonetto) [reply core/src/test/java/org/opensearch/sql/analysis/AnalyzerTest.java:375]: unit tests are better

**Review by acarbonetto — APPROVED** [review]

File: sql/src/main/antlr/OpenSearchSQLParser.g4 [hunk 19]
```diff
@@ -364,6 +365,22 @@ datetimePart
...
+nestedField
+    : nestedIdent DOT nestedIdent (DOT nestedIdent)*
+    | stringLiteral
+    ;
+
+nestedPath
+    : nestedIdent (DOT nestedIdent)*
+    | stringLiteral
+    ;
+
```
Comment (GabeFernandez310) [inline sql/src/main/antlr/OpenSearchSQLParser.g4:383]: Not sure if it's an issue for this PR (I think it's more related to the duplicate columns POC), but I was able to have the following query fail.

```
opensearchsql> SELECT nested(message.info) AS a, nested(message.info) AS b FROM nested;
{'reason': 'Invalid SQL query', 'details': 'Multiple entries with same key: message.info="a" and message.info="a"', 'type': 'IllegalArgumentException'}
```

Switching the nested field works though.

```
opensearchsql> SELECT nested(message.author) AS a, nested(message.info) AS b FROM nested;
fetched rows / total rows = 8/8
+------------------+----------------+
| message.author   | message.info   |
|------------------+----------------|
| e                | a              |
| f                | b              |
| g                | c              |
| h                | c              |
| i                | c              |
| h                | a              |
| i                | a              |
| zz               | zz             |
+------------------+----------------+
```

Just wondering if it's an issue because for non-nested queries, the same field given a different alias removes the error returned in the first query 

```
opensearchsql> SELECT someField AS a, someField AS b FROM nested;
fetched rows / total rows = 5/5
+-----+-----+
| a   | b   |
|-----+-----|
| b   | b   |
| a   | a   |
| a   | a   |
| b   | b   |
| a   | a   |
+-----+-----+
```

File: core/src/main/java/org/opensearch/sql/planner/physical/UnnestOperator.java (hunk 17 above)
Comment (forestmvey) [reply core/src/main/java/org/opensearch/sql/planner/physical/UnnestOperator.java:163]: We can but we will need another function that is almost the same as `getNested`.

File: core/src/main/java/org/opensearch/sql/planner/physical/UnnestOperator.java [hunk 20]
```diff
@@ -0,0 +1,252 @@
...
+      result.clear();
+      nonNestedFields.clear();
+
+      ExprValue inputValue = input.next();
+      generateNonNestedFieldsMap(inputValue);
+      for (String field : fields) {
+        result = flatten(field, inputValue, result, true);
+      }
+
+      if (result.isEmpty()) {
```
[the remark this answers is not in the document]
Comment (forestmvey) [reply core/src/main/java/org/opensearch/sql/planner/physical/UnnestOperator.java]: Shouldn't happen, fixed in [5688f5c47b4c7902cb2636c58d9d2995feb0a8ff](https://github.com/Bit-Quill/opensearch-project-sql/pull/243/commits/5688f5c47b4c7902cb2636c58d9d2995feb0a8ff)

**Comment by forestmvey** [comment]
> Does documentation/doctests need to be added?

Added doctest in [10ee84d14f28b1c7e5bfa92a76e92cf9c2d74579](https://github.com/Bit-Quill/opensearch-project-sql/pull/243/commits/10ee84d14f28b1c7e5bfa92a76e92cf9c2d74579)

File: core/src/main/java/org/opensearch/sql/planner/physical/UnnestOperator.java (hunk 16 above)
Comment (forestmvey) [reply core/src/main/java/org/opensearch/sql/planner/physical/UnnestOperator.java]: Fixed in [a574cb6d78483e1b329d5db5a9fe57d03ffa4191](https://github.com/Bit-Quill/opensearch-project-sql/pull/243/commits/a574cb6d78483e1b329d5db5a9fe57d03ffa4191)

File: core/src/main/java/org/opensearch/sql/planner/physical/UnnestOperator.java [hunk 21]
```diff
@@ -0,0 +1,252 @@
...
+   * Generate list of non-nested fields that are in inputMap, but not in the member variable
+   * fields list.
+   * @param inputMap : Row to parse non-nested fields.
+   */
+  public void generateNonNestedFieldsMap(ExprValue inputMap) {
+
+    for (Map.Entry<String, ExprValue> inputField : inputMap.tupleValue().entrySet()) {
+      boolean foundNestedField = false;
+      for (String field : this.fields) {
+        if ((field).split("\\.")[0].equalsIgnoreCase(inputField.getKey())) {
```
[the remark this answers is not in the document]
Comment (forestmvey) [reply core/src/main/java/org/opensearch/sql/planner/physical/UnnestOperator.java]: Thanks, fixed in [5688f5c47b4c7902cb2636c58d9d2995feb0a8ff](https://github.com/Bit-Quill/opensearch-project-sql/pull/243/commits/5688f5c47b4c7902cb2636c58d9d2995feb0a8ff)

**Review by GabeFernandez310 — APPROVED** [review]

File: sql/src/main/antlr/OpenSearchSQLParser.g4 (hunk 19 above)
Comment (forestmvey) [reply sql/src/main/antlr/OpenSearchSQLParser.g4:383]: Nice catch, seems there was a bug with Aliases. It has been resolved in [d7055a9b7f91ecb7a5e03c40a1631e005305760d](https://github.com/Bit-Quill/opensearch-project-sql/pull/243/commits/d7055a9b7f91ecb7a5e03c40a1631e005305760d)

**Review by GabeFernandez310 — APPROVED** [review]

File: core/src/test/java/org/opensearch/sql/analysis/AnalyzerTest.java (hunk 18 above)
[the remark this answers is not in the document]
Comment (forestmvey) [reply core/src/test/java/org/opensearch/sql/analysis/AnalyzerTest.java:375]: Added a UT covering deep nested!

File: core/src/main/java/org/opensearch/sql/planner/physical/UnnestOperator.java [hunk 22]
```diff
@@ -0,0 +1,291 @@
...
+  public UnnestOperator(PhysicalPlan input, List<Map<String, ReferenceExpression>> fields) {
+    this.input = input;
+    this.fields = fields.stream()
+        .map(m -> m.get("field").toString())
+        .collect(Collectors.toSet());
+    this.groupedPathsAndFields = fields.stream().collect(
+        Collectors.groupingBy(
+            m -> m.get("path").toString(),
+            mapping(
+                m -> m.get("field").toString(),
```
Comment (Yury-Fridlyand) [inline core/src/main/java/org/opensearch/sql/planner/physical/UnnestOperator.java:56]: Maybe use constants?
Maybe use class fields? `m.getField()` and `m.getPath()`?

File: core/src/main/java/org/opensearch/sql/planner/physical/UnnestOperator.java [hunk 23]
```diff
@@ -0,0 +1,291 @@
...
+import org.opensearch.sql.expression.ReferenceExpression;
+
+@EqualsAndHashCode(callSuper = false)
+public class UnnestOperator extends PhysicalPlan {
+  @Getter
+  private final PhysicalPlan input;
+  @Getter
+  private final Set<String> fields; // Needs to be a Set to match legacy implementation
+  @Getter
+  private final Map<String, List<String>> groupedPathsAndFields;
```
Comment (Yury-Fridlyand) [inline core/src/main/java/org/opensearch/sql/planner/physical/UnnestOperator.java:36]: Do you need to store both? You can calculate second field from first when you need.
Use `LinkedHashMap` to keep the insert order.

File: core/src/main/java/org/opensearch/sql/planner/physical/UnnestOperator.java [hunk 24]
```diff
@@ -0,0 +1,291 @@
...
+
+      ExprValue inputValue = input.next();
+      generateNonNestedFieldsMap(inputValue);
+      for (String field : fields) {
+        result = flatten(field, inputValue, result, true);
+      }
+
+      for (String nonNestedField : nonNestedFields) {
+        result = flatten(nonNestedField, inputValue, result, false);
+      }
```
Comment (Yury-Fridlyand) [inline core/src/main/java/org/opensearch/sql/planner/physical/UnnestOperator.java:109]: Is it safe to overwrite `result` in the loop? Maybe leave a comment

File: core/src/main/java/org/opensearch/sql/planner/physical/UnnestOperator.java [hunk 25]
```diff
@@ -0,0 +1,291 @@
...
+  }
+
+  /**
+   * Check if newMap field has any sharing paths in prevMap.
+   * @param newMap : New map to add to result set.
+   * @return : true if there is already a field added to result set with same path.
+   */
+  boolean containSamePath(Map<String, ExprValue> newMap) {
+    String newKey = newMap.keySet().iterator().next();
+    Map<String, ExprValue> resultMap = this.result.iterator().next();
```
Comment (Yury-Fridlyand) [inline core/src/main/java/org/opensearch/sql/planner/physical/UnnestOperator.java:225]: You check only first entry in `result`. Why?

File: docs/user/dql/functions.rst [hunk 26]
```diff
@@ -4299,6 +4299,28 @@ Another example to show how to set custom values for the optional parameters::
     | tEsT wIlDcArD sensitive cases             |
     +-------------------------------------------+
 
+NESTED
+------------
+
+Description
+>>>>>>>>>>>
+
+``nested(field | [field, path] | [path, condition])``
```
Comment (Yury-Fridlyand) [inline docs/user/dql/functions.rst:4308]: Give examples for each syntax option, please.

File: docs/user/dql/functions.rst [hunk 27]
```diff
@@ -4299,6 +4299,28 @@ Another example to show how to set custom values for the optional parameters::
...
+
+The ``nested`` function maps to the ``nested`` query used in search engine. It returns nested field types in documents that match the provided specified field(s).
+The ``field`` and ``path`` parameters are valid parameters for a nested function used in the ``SELECT`` clause.
```
Comment (Yury-Fridlyand) [inline docs/user/dql/functions.rst:4311]: Maybe reword
> The ``field`` and ``path`` parameters are valid parameters ...

What about `condition`? If it is not supported, remove it from syntax.
I propose mention that `nested` can be used in `SELECT` only in a separate sentence.

File: docs/user/dql/functions.rst [hunk 28]
```diff
@@ -4299,6 +4299,28 @@ Another example to show how to set custom values for the optional parameters::
     | tEsT wIlDcArD sensitive cases             |
     +-------------------------------------------+
 
+NESTED
+------------
```
Comment (Yury-Fridlyand) [inline docs/user/dql/functions.rst:4303]:
```suggestion

NESTED
------
```

Comment (Yury-Fridlyand) [reply docs/user/dql/functions.rst:4303]: Consider adding a header (section) for `nested`

File: doctest/test_mapping/nested_objects.json [hunk 29]
```diff
@@ -0,0 +1,47 @@
...
+      "message": {
+        "type": "nested",
+        "properties": {
+          "info": {
+            "type": "keyword",
+            "index": "true"
+          },
+          "author": {
+            "type": "keyword",
+            "fields": {
```
Comment (Yury-Fridlyand) [inline doctest/test_mapping/nested_objects.json:13]: It is possible to have `keyword` with `fields`? How it works?

File: integ-test/src/test/java/org/opensearch/sql/sql/NestedIT.java (hunk 3 above)
Comment (Yury-Fridlyand) [reply integ-test/src/test/java/org/opensearch/sql/sql/NestedIT.java:116]: Agree, but please add tests 
1. with condition
2. with null and missing cases
For example, index:
```
doc_1: "message": { "info" : "something" }
doc_2: "message": { "info" : "another thing" }
doc_3: "message": { "info" : null }
doc_4: "message": { }
doc_5: // no message
```
Maybe combine this with multiple `nested` calls and/or deeper nesting to prove that code in `UnnestOperator` is solid.

File: sql/src/main/antlr/OpenSearchSQLParser.g4 [hunk 30]
```diff
@@ -369,6 +370,22 @@ datetimePart
     | complexDateTimePart
     ;
 
+nestedFunction
+    : NESTED LR_BRACKET
+    (nestedField | nestedField COMMA nestedPath)
```
Comment (Yury-Fridlyand) [inline sql/src/main/antlr/OpenSearchSQLParser.g4:375]: Maybe
```suggestion
    nestedField | (nestedField COMMA nestedPath)
```

File: core/src/main/java/org/opensearch/sql/analysis/Analyzer.java [hunk 31]
```diff
@@ -359,6 +359,14 @@ public LogicalPlan visitProject(Project node, AnalysisContext context) {
     List<NamedExpression> namedExpressions =
         selectExpressionAnalyzer.analyze(node.getProjectList(), context,
             new ExpressionReferenceOptimizer(expressionAnalyzer.getRepository(), child));
+
+    for (UnresolvedExpression expr : node.getProjectList()) {
+      NestedAnalyzer nestedAnalyzer = new NestedAnalyzer(
+          namedExpressions, expressionAnalyzer, child
+      );
+      child = nestedAnalyzer.analyze(expr, context);
```
[the remark this answers is not in the document]
Comment (forestmvey) [reply core/src/main/java/org/opensearch/sql/analysis/Analyzer.java:367]: Don't believe we have any requirement for `nested` to work with added V2 functionality like `highlight`. Goal is legacy deprecation and we want to abandon the NLPChina implementation for a better standard like PartiQL. I can pushDown `highlight` and `nested`, I'm not sure what a valid query would be though?

File: core/src/main/java/org/opensearch/sql/analysis/NestedAnalyzer.java [hunk 32]
```diff
@@ -0,0 +1,69 @@
...
+            "field", nestedField,
+            "path", (ReferenceExpression)expressionAnalyzer.analyze(expressions.get(1), context)
+        );
+      } else {
+        args = Map.of(
+            "field", (ReferenceExpression)expressionAnalyzer.analyze(expressions.get(0), context),
+            "path", generatePath(nestedField.toString())
+        );
+      }
+      return new LogicalNested(child, List.of(args), namedExpressions);
```
[the remark this answers is not in the document]
Comment (forestmvey) [reply core/src/main/java/org/opensearch/sql/analysis/NestedAnalyzer.java:61]: See [MergeNestedAndNested](https://github.com/Bit-Quill/opensearch-project-sql/blob/26aa7c6318852698033fddbaa9a359f3c968d3a4/core/src/main/java/org/opensearch/sql/planner/optimizer/rule/MergeNestedAndNested.java)

File: core/src/main/java/org/opensearch/sql/analysis/NestedAnalyzer.java [hunk 33]
```diff
@@ -0,0 +1,69 @@
...
+    }
+    return null;
```
[the remark this answers is not in the document]
Comment (forestmvey) [reply core/src/main/java/org/opensearch/sql/analysis/NestedAnalyzer.java:63]: No the `NestedAnalyzer` will visit the entire project list so we cannot return child here.

File: doctest/test_mapping/nested_objects.json (hunk 29 above)
Comment (forestmvey) [reply doctest/test_mapping/nested_objects.json:13]: Not entirely sure, this was taken from [nested_type_index_mapping.json](https://github.com/opensearch-project/sql/blob/main/integ-test/src/test/resources/indexDefinitions/nested_type_index_mapping.json)

File: sql/src/main/antlr/OpenSearchSQLParser.g4 (hunk 30 above)
Comment (forestmvey) [reply sql/src/main/antlr/OpenSearchSQLParser.g4:375]: This breaks all nested functions in the parser.

File: core/src/main/java/org/opensearch/sql/expression/nested/NestedFunction.java [hunk 34]
```diff
@@ -0,0 +1,38 @@
...
+   * @param arguments a list of expressions
+   */
+  public NestedFunction(FunctionName functionName, List<Expression> arguments) {
+    super(functionName, arguments);
+    this.arguments = arguments;
+  }
+
+  @Override
+  public ExprValue valueOf(Environment<Expression, ExprValue> valueEnv) {
+    return valueEnv.resolve(this.arguments.get(0));
```
[the remark this answers is not in the document]
Comment (forestmvey) [reply core/src/main/java/org/opensearch/sql/expression/nested/NestedFunction.java:31]: I'll I'll add a comment with a GH issue to current array support in V2 engine!

File: core/src/main/java/org/opensearch/sql/planner/physical/UnnestOperator.java (hunk 25 above)
Comment (forestmvey) [reply core/src/main/java/org/opensearch/sql/planner/physical/UnnestOperator.java:238]: In the event we are adding a nested field with another fields same path to the result set, we don't want to do a cross join on them. So we check which paths are already added to the results map to compare with the new map being added. Even if a previous item was `null` or `missing` their key will still be in the results map. Therefore we only care to check the first entry of `result`.

File: core/src/main/java/org/opensearch/sql/planner/physical/UnnestOperator.java (hunk 23 above)
Comment (forestmvey) [reply core/src/main/java/org/opensearch/sql/planner/physical/UnnestOperator.java:36]: I think this is the easiest way to accomplish the comparison in `containSamePath` function. When adding a new key to the result map we need to check: 

- Is there any matching keys with the same path.
- Is a matching key with the same path already in the result map.

Rather than calculating the path for each field already in the results map, and each field being added to the results map we maintain this member variable.

File: core/src/main/java/org/opensearch/sql/planner/physical/UnnestOperator.java (hunk 24 above)
Comment (forestmvey) [reply core/src/main/java/org/opensearch/sql/planner/physical/UnnestOperator.java:109]: Added comments for each `flatten` function. Do you see something that isn't safe?

File: core/src/main/java/org/opensearch/sql/analysis/Analyzer.java (hunk 31 above)
[the remark this answers is not in the document]
Comment (Yury-Fridlyand) [reply core/src/main/java/org/opensearch/sql/analysis/Analyzer.java:367]: `select nested(...), highlight(...) from ... where match(...)`
Can you try this?

**Review by Yury-Fridlyand — APPROVED** [review]

File: core/src/main/java/org/opensearch/sql/analysis/Analyzer.java (hunk 31 above)
[the remark this answers is not in the document]
Comment (forestmvey) [reply core/src/main/java/org/opensearch/sql/analysis/Analyzer.java:367]: Issue here for tracking.
[1488](https://github.com/opensearch-project/sql/issues/1488)
